presto-elasticsearch/src/test/java/com/facebook/presto/elasticsearch/ElasticsearchLoader.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.elasticsearch;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.client.Column;
+import com.facebook.presto.client.QueryData;
+import com.facebook.presto.client.QueryStatusInfo;
+import com.facebook.presto.server.testing.TestingPrestoServer;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.Varchars;
+import com.facebook.presto.tests.AbstractTestingPrestoClient;
+import com.facebook.presto.tests.ResultsSession;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.type.DateType.DATE;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.elasticsearch.common.xcontent.XContentType.JSON;
+
+public class ElasticsearchLoader
+        extends AbstractTestingPrestoClient<Void>
+{
+    private final String tableName;
+    private final Client client;
+
+    public ElasticsearchLoader(Client client,
+            String tableName,
+            TestingPrestoServer prestoServer,
+            Session defaultSession)
+    {
+        super(prestoServer, defaultSession);
+
+        this.tableName = requireNonNull(tableName, "tableName is null");
+        this.client = requireNonNull(client, "client is null");
+    }
+
+    @Override
+    public ResultsSession<Void> getResultSession(Session session)
+    {
+        requireNonNull(session, "session is null");
+        return new ElasticsearchLoadingSession();
+    }
+
+    private class ElasticsearchLoadingSession
+            implements ResultsSession<Void>
+    {
+        private final AtomicReference<List<Type>> types = new AtomicReference<>();
```

Comment:
Why do you need an `AtomicReference`?

Reply:
my understanding is, types guarantee multiple sessions could be live concurrently to fetch results. Fix me if I am wrong